### DISCUSSION TITLE
resolveComponent [Vue warn]: Failed to resolve component:

### DISCUSSION THREAD

**Discussion by cpt-n3mo** [nuxt/nuxt#17411] [Questions] [answered]
Im trying to load a component list from a database, so that i can dynamically render different components.

```javascript
const getComponent =  (name) => {
    return resolveComponent(name)
}
```
does not seem to work as expected as it cant find the component. saying..
`[Vue warn]: Failed to resolve component: componentName`

when i use the same component name directly inside `resolveComponent('componentName')` it works just fine 

am i missing something here..?

regards

**Comment by danielroe** [comment, marked as answer]
Yes. Nuxt does some magic to auto-import components that are resolved like `resolveComponent('componentName')`. If you want it to work fully programmatically, you'll need to set `components.global` to true so that Vue can resolve them.

**Reply by settel** [reply]
Came across same issue and eventually found this thread. The magical behaviour isn't obvious from reading https://v3.nuxtjs.org/guide/directory-structure/components/#dynamic-components . 

Considering snippet below. To me, there really should be no difference between "1:" and "2:" but in practice there is: 
* "1:" throws a warning on server but renders on client with "Hydration children mismatch in `<div>`" warning
* "2:" doesn't even render component but outputs a HTML tag

I'd rather not have magical surprises happen but prefer to have an obvious API.

<hr>

app.vue:
```vue
<template>
  <div>
    <div>1: <component :is="DynamicComponent1" /></div>
    <div>2: <component :is="DynamicComponent2" /></div>
  </div>
</template>

<script setup>
const DynamicComponent1 = resolveComponent('MyComponent')
const DynamicComponent2 = resolveComponent(`MyComponent`)
</script>
```
components/MyComponent.vue
```vue
<template>
  This is MyComponent.vue
</template>
```

output (excerpt from final DOM):
```html
<div>1: This is MyComponent.vue</div>
<div>2: <mycomponent></mycomponent></div>
```
additional information:
```bash
 ~ $ yarn list --pattern nuxt3
yarn list v1.22.18
└─ nuxt3@3.0.0-27497942.c519cee
Done in 0.26s.
```

**Reply by settel** [reply]
with danielroe's solution, the aforementioned snippet renders in client but still throws warning on browser console:


nuxt.config.ts
```typescript
export default defineNuxtConfig({
  components: {
    global: true,
    dirs: ['~/components'],
  },
})
```

browser console
```
[vite] connecting... [client.ts:16:8](http://localhost:3000/src/client/client.ts)
[vite] connected. [client.ts:53:14](http://localhost:3000/src/client/client.ts)
<Suspense> is an experimental feature and its API will likely change. [runtime-core.esm-bundler.js:1326:47](http://localhost:3000/_nuxt/node_modules/@vue/runtime-core/dist/runtime-core.esm-bundler.js)
[Vue warn]: Hydration text mismatch:
- Client: "1:  This is MyComponent.vue "
- Server: "1: " 
  at <App key=1 > 
  at <NuxtRoot> [runtime-core.esm-bundler.js:38:16](http://localhost:3000/_nuxt/node_modules/@vue/runtime-core/dist/runtime-core.esm-bundler.js)
[Vue warn]: Hydration children mismatch in <div>: server rendered element contains fewer child nodes than client vdom. 
  at <App key=1 > 
  at <NuxtRoot> [runtime-core.esm-bundler.js:38:16](http://localhost:3000/_nuxt/node_modules/@vue/runtime-core/dist/runtime-core.esm-bundler.js)
[Vue warn]: Hydration text mismatch:
- Client: "2:  This is MyComponent.vue "
- Server: "2: " 
  at <App key=1 > 
  at <NuxtRoot> [runtime-core.esm-bundler.js:38:16](http://localhost:3000/_nuxt/node_modules/@vue/runtime-core/dist/runtime-core.esm-bundler.js)
[Vue warn]: Hydration children mismatch in <div>: server rendered element contains fewer child nodes than client vdom. 
  at <App key=1 > 
  at <NuxtRoot> [runtime-core.esm-bundler.js:38:16](http://localhost:3000/_nuxt/node_modules/@vue/runtime-core/dist/runtime-core.esm-bundler.js)
Hydration completed but contains mismatches. [runtime-core.esm-bundler.js:4058:20](http://localhost:3000/_nuxt/node_modules/@vue/runtime-core/dist/runtime-core.esm-bundler.js)
    NuxtJS 4
```

Should I file a bug?

**Reply by ilyaDegtyarenko** [reply]
> Yes. Nuxt does some magic to auto-import components that are resolved like `resolveComponent('componentName')`. If you want it to work fully programmatically, you'll need to set `components.global` to true so that Vue can resolve them.

Is there a more convenient solution?
In my case, I use a button component in which I can use the icon component (I pass the name of icon component in props). And when I add a new icon as a component, I need to remember that I need to add IT to the button component, because it won't work.

**Reply by ilyaDegtyarenko** [reply]
Now I'm used slot

**Reply by vitalijalbu** [reply]
Does it work also with vue as well? Cause I’m using nuxt ui + vue
